test/parallel/test-stream-pipeline.js

```diff
@@ -483,3 +490,26 @@ const { promisify } = require('util');
     { code: 'ERR_INVALID_CALLBACK' }
   );
 }
+
+{
+  const server = http.Server(function(req, res) {
+    res.write('asd');
+  });
+  server.listen(0, function() {
+    http.get({ port: this.address().port }, (res) => {
+      const stream = new PassThrough();
+
+      stream.on('error', common.mustCall());
+
+      pipeline(
+        res,
+        stream,
+        common.mustCall((err) => {
+          server.close();
```

Comment:
Does it make sense to also validate the error here?

Reply:
I agree. At least that `err` is truthy.

Reply:
Fixed. Unfortunately we get an unexpected/different error due to the problems discussed in https://github.com/nodejs/node/issues/31060. So I made it just check for truthyness.

Reply:
I can't understand how this is related to #31090. AFAIK none of the `_destroy()` implementation is async.

It seems a regression on master to me

```js
const stream = require('stream');

const data = Buffer.alloc(1024);

{
  const readable = new stream.Readable({
    read() {
      this.push(data);
    }
  });

  const writable = new stream.Writable({
    write(chunk, encoding, callback) {
      callback();
    }
  });

  writable.on('error', console.error);

  readable.pipe(writable);
  writable.destroy(new Error('Oops'));
}

{
  const readable = new stream.Readable({
    read() {
      this.push(data);
    }
  });

  const writable = new stream.Writable({
    write(chunk, encoding, callback) {
      callback();
    }
  });

  stream.pipeline(readable, writable, console.error);

  writable.destroy(new Error('Oops'));
}
```

This prints

```
Error: First
    at Object.<anonymous> (/Users/luigi/Desktop/pipe.js:23:20)
    at Module._compile (internal/modules/cjs/loader.js:1139:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1159:10)
    at Module.load (internal/modules/cjs/loader.js:988:32)
    at Function.Module._load (internal/modules/cjs/loader.js:896:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
Error: First
    at Object.<anonymous> (/Users/luigi/Desktop/pipe.js:41:20)
    at Module._compile (internal/modules/cjs/loader.js:1139:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1159:10)
    at Module.load (internal/modules/cjs/loader.js:988:32)
    at Function.Module._load (internal/modules/cjs/loader.js:896:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
```

on Node.js 13.5.0 and

```
Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at Writable.write (_stream_writable.js:321:17)
    at Readable.ondata (_stream_readable.js:779:22)
    at Readable.emit (events.js:320:20)
    at Readable.read (_stream_readable.js:579:10)
    at flow (_stream_readable.js:1052:34)
    at resume_ (_stream_readable.js:1033:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'ERR_STREAM_DESTROYED'
}
Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at Writable.write (_stream_writable.js:321:17)
    at Readable.ondata (_stream_readable.js:779:22)
    at Readable.emit (events.js:320:20)
    at Readable.read (_stream_readable.js:579:10)
    at flow (_stream_readable.js:1052:34)
    at resume_ (_stream_readable.js:1033:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'ERR_STREAM_DESTROYED'
}
```

on Node.js master

Reply:
Bisecting points to 67ed526ab03c3e9b2f41290b7cd6e9c2e7d9b4dd.

Reply:
@lpinca: The stream is `.destroyed` but `pipe` will continue writing causing a sync `ERR_STREAM_DESTOYRED` which will gain precedence.

Reply:
Simpler example:

```js
const { Readable, Writable } = require('stream');

const r = new Readable({
  read () {
    this.push('asd');
  }
});

const w = new Writable({
  write (chunk, encoding, callback) {
    process.nextTick(callback);
  }
})

r
  .pipe(w)
  .on('error', (err) => {
    console.log(err); // ERR_STREAM_DESTROYED is first
  });

w.destroy(new Error('kaboom'));
```

Reply:
It does not happen with Node.js 13.5.0

Reply:
@lpinca: You are right. It doesn't have any async in destruction but since `write` emits the error sync (fixed in https://github.com/nodejs/node/pull/29744) the problem is the same as if the writable had an async `_destroy` (in which case we are back here).

Reply:
> It does not happen with Node.js 13.5.0

Yes, looks like Node 13.5.0 had another bug which would hide this.

Reply:
Yes, again `git bisect` points to 67ed526ab03c3e9b2f41290b7cd6e9c2e7d9b4dd where the `errorEmitted` guard was moved to the same tick where the error is emitted.

Reply:
> the problem is the same as if the writable had an async `_destroy` (in which case we are back here with the same "problem")

I disagree on this. The problem is not the same. In this case there is no ambiguity. `_destroy()` callback is called with an error (but the error is emitted later) **before** `writable.write()` is called. The `writable.write()` error should not be emitted.

Reply:
> In this case there is no ambiguity.

Yes, if the write ERR_STREAM_DESTROYED error was async (as it should be) there would be no ambiguity.

Reply:
Even if it is not. In this case `writable.destroy()` is fully completed and with an error before `writable.write()` occurs.

Reply:
write is called before the error is emitted since the error is emitted in a tick later

Reply:
Yes, that's why the guard was set on destroy complete (before https://github.com/nodejs/node/commit/67ed526ab03c3e9b2f41290b7cd6e9c2e7d9b4dd) instead of the actual tick where the error is emitted, I guess.

Reply:
So ignore destroy with error after _destroy has completed. Is that another option to consider?

Reply:
If that completed `_destroy()` completed with an error yes, totally.

Reply:
Cool. I’ll add it to the proposals list.

Reply:
How can this thread be resolved? We need a fix with the bug this PR aims to solve for node v13, or better revert the few commits that caused the problem in the first place.

Reply:
I don't think this specific thread needs to be urgently resolved. This PR resolved the fundamental issue.

Reply:
Agreed, this discussion should not block the PR. The issue discussed here is caused by a semver-major change that will not be included in v13.x